internal/api/mfa.go

```diff
@@ -205,6 +367,128 @@ func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	})
 }
 
+func (a *API) verifySMSFactor(w http.ResponseWriter, r *http.Request, user *models.User, factor *models.Factor, params *VerifyFactorParams) error {
+	ctx := r.Context()
+	config := a.config
+	db := a.db.WithContext(ctx)
+	if !config.MFA.SMS.VerifyEnabled {
+		return unprocessableEntityError(ErrorCodeMFASMSVerifyDisabled, "MFA verification is disabled for sms factors")
+	}
+	currentIP := utilities.GetIPAddress(r)
+
+	if !factor.IsOwnedBy(user) {
+		return internalServerError(InvalidFactorOwnerErrorMessage)
+	}
+
+	challenge, err := models.FindChallengeByID(db, params.ChallengeID)
+	if err != nil && models.IsNotFoundError(err) {
+		return notFoundError(ErrorCodeMFAFactorNotFound, "MFA factor with the provided challenge ID not found")
+	} else if err != nil {
+		return internalServerError("Database error finding Challenge").WithInternalError(err)
+	}
+
+	if challenge.VerifiedAt != nil || challenge.IPAddress != currentIP {
+		return unprocessableEntityError(ErrorCodeMFAIPAddressMismatch, "Challenge and verify IP addresses mismatch")
+	}
+
+	if challenge.HasExpired(config.MFA.ChallengeExpiryDuration) {
+		if err := db.Destroy(challenge); err != nil {
+			return internalServerError("Database error deleting challenge").WithInternalError(err)
+		}
+		return unprocessableEntityError(ErrorCodeMFAChallengeExpired, "MFA challenge %v has expired, verify against another challenge or create a new challenge.", challenge.ID)
+	}
+	otpCode, shouldReEncrypt, err := challenge.GetOtpCode(config.Security.DBEncryption.DecryptionKeys, config.Security.DBEncryption.Encrypt, config.Security.DBEncryption.EncryptionKeyID)
+	if err != nil {
+		return internalServerError("Database error verifying MFA TOTP secret").WithInternalError(err)
+	}
+	valid := subtle.ConstantTimeCompare([]byte(otpCode), []byte(params.Code)) == 1
```

Comment:
some weird decryption error going on here - i think it's still decrypting the otp stored in the db even though it was stored as plaintext

Reply:
The decryption was likely related to [this line](https://github.com/supabase/auth/pull/1668/commits/cff5e9d52ed4a20aeb03aa6034e08316e0e8d4bd#diff-3d75bd728617348675d2aaf07152fa3579f00f57f6f943a20cb1c25314af7e54R78) it was previously returning the `ChallengeID` - tested recently and it should work now